miner/scroll_worker.go

```diff
@@ -316,133 +352,152 @@ func (w *worker) mainLoop() {
-		return
+		return fmt.Errorf("failed to prepare header for mining: %w", err)
 	}
 	prepareTimer.UpdateSince(prepareStart)
 
-	// If we are care about TheDAO hard-fork check whether to override the extra-data or not
-	if daoBlock := w.chainConfig.DAOForkBlock; daoBlock != nil {
-		// Check whether the block is among the fork extra-override range
-		limit := new(big.Int).Add(daoBlock, params.DAOForkExtraRange)
-		if header.Number.Cmp(daoBlock) >= 0 && header.Number.Cmp(limit) < 0 {
-			// Depending whether we support or oppose the fork, override differently
-			if w.chainConfig.DAOForkSupport {
-				header.Extra = common.CopyBytes(params.DAOForkBlockExtra)
-			} else if bytes.Equal(header.Extra, params.DAOForkBlockExtra) {
-				header.Extra = []byte{} // If miner opposes, don't let it use the reserved extra-data
-			}
-		}
+	var nextL1MsgIndex uint64
+	if dbVal := rawdb.ReadFirstQueueIndexNotInL2Block(w.eth.ChainDb(), header.ParentHash); dbVal != nil {
+		nextL1MsgIndex = *dbVal
 	}
 
-	parentState, err := w.chain.StateAt(parent.Root())
+	vmConfig := *w.chain.GetVMConfig()
+	cccLogger := ccc.NewLogger(types.RowConsumptionLimit)
+	vmConfig.Debug = true
+	vmConfig.Tracer = cccLogger
+
+	deadline := time.Unix(int64(header.Time), 0)
+	if w.chainConfig.Clique != nil && w.chainConfig.Clique.RelaxedPeriod {
+		// clique with relaxed period uses time.Now() as the header.Time, calculate the deadline
+		deadline = time.Unix(int64(header.Time+w.chainConfig.Clique.Period), 0)
+	}
+
+	w.current = &work{
+		deadlineTimer:  time.NewTimer(time.Until(deadline)),
+		cccLogger:      cccLogger,
+		vmConfig:       vmConfig,
+		header:         header,
+		state:          parentState,
+		txs:            types.Transactions{},
+		receipts:       types.Receipts{},
+		coalescedLogs:  []*types.Log{},
+		gasPool:        new(core.GasPool).AddGas(header.GasLimit),
+		nextL1MsgIndex: nextL1MsgIndex,
+		reorgReason:    reorgReason,
+	}
+	return nil
+}
+
+// tryCommitNewWork
+func (w *worker) tryCommitNewWork(now time.Time, parent common.Hash, reorgReason error) (common.Hash, error) {
+	err := w.newWork(now, parent, reorgReason)
 	if err != nil {
-		log.Error("failed to fetch parent state", "err", err)
-		return
+		return common.Hash{}, fmt.Errorf("failed creating new work: %w", err)
 	}
 
-	// Apply special state transition at Curie block
-	if w.chainConfig.CurieBlock != nil && w.chainConfig.CurieBlock.Cmp(header.Number) == 0 {
-		misc.ApplyCurieHardFork(parentState)
+	shouldCommit, err := w.handleForks()
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("failed handling forks: %w", err)
+	}
 
-		var nextL1MsgIndex uint64
-		if dbVal := rawdb.ReadFirstQueueIndexNotInL2Block(w.eth.ChainDb(), header.ParentHash); dbVal != nil {
-			nextL1MsgIndex = *dbVal
-		}
+	// check if we are reorging
+	reorging := w.chain.GetBlockByNumber(w.current.header.Number.Uint64()) != nil
+	if !shouldCommit && reorging {
+		shouldCommit, err = w.processReorgedTxns(w.current.reorgReason)
```

Comment:
Why update `shouldCommit` here? Seems like it's always true

Reply:
We might decide to let worker process txn pool as well after handling reorged txns.